### ISSUE TITLE
Trying to upgrade to newest version

### ISSUE THREAD

**Issue by jimcrist** [josephlacey/com.jlacey.electoral#33]
I started with V1 in 2016 and when I went to V2 I had to delete the district data from the database and recreate it all.  I just upgraded from a two year old version of CiviCRM to the newest version and it said the Electoral API was missing files.  I wanted the newest version of the Electoral API anyway, so I uninstalled the non-functioning version, copied your newest files from https://github.com/josephlacey/com.jlacey.electoral and tried to install it. Now I'm getting DB Error: already exists
So it remains uninstalled and I'm stuck.  I suppose I never should have uninstalled it.  Just copied the program files in. Rats.

Please advise...thanks!
Jim Crist

**Comment by josephlacey** [comment]
Hi @jimcrist,

If you can get more details on exactly what the install is complaining about, I could give some advise here.  It's likely either trying to recreate the custom data fields or the options values that already exist.  We don't remove either of those in case folks want to keep the data, but there's also not a way to remove those from the UI.  It should be possible to remove a file or part of a file to get the install completed.

Joseph

**Comment by jimcrist** [comment]
OK, so I went into CiviCRM Administer > System Settings > Debugging and Error Handling and turned on Debugging and also Backtrace.  This looks reminiscent to https://github.com/josephlacey/com.jlacey.electoral/issues/23
Down below that is what shows up in the ConfigAndLog file.  Please advise...thanks!

CiviCRM_API3_Exception: "DB Error: already exists"
#0 .../wp-content/plugins/civicrm/civicrm/CRM/Admin/Form/Extensions.php(173): civicrm_api3("Extension", "install", (Array:2))
#1 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Form.php(526): CRM_Admin_Form_Extensions->postProcess()
#2 .../wp-content/plugins/civicrm/civicrm/CRM/Core/StateMachine.php(144): CRM_Core_Form->mainProcess()
#3 .../wp-content/plugins/civicrm/civicrm/CRM/Core/QuickForm/Action/Next.php(43): CRM_Core_StateMachine->perform(Object(CRM_Admin_Form_Extensions), "next", "Next")
#4 .../wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Controller.php(203): CRM_Core_QuickForm_Action_Next->perform(Object(CRM_Admin_Form_Extensions), "next")
#5 .../wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Page.php(103): HTML_QuickForm_Controller->handle(Object(CRM_Admin_Form_Extensions), "next")
#6 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Controller.php(347): HTML_QuickForm_Page->handle("next")
#7 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Page/Basic.php(395): CRM_Core_Controller->run()
#8 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Page/Basic.php(140): CRM_Core_Page_Basic->edit(1, NULL)
#9 .../wp-content/plugins/civicrm/civicrm/CRM/Admin/Page/Extensions.php(105): CRM_Core_Page_Basic->run()
#10 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(313): CRM_Admin_Page_Extensions->run((Array:3), NULL)
#11 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(69): CRM_Core_Invoke::runItem((Array:15))
#12 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(36): CRM_Core_Invoke::_invoke((Array:3))
#13 .../wp-content/plugins/civicrm/civicrm.php(1172): CRM_Core_Invoke::invoke((Array:3))
#14 .../wp-includes/class-wp-hook.php(292): CiviCRM_For_WordPress->invoke("")
#15 .../wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters("", (Array:1))
#16 .../wp-includes/plugin.php(484): WP_Hook->do_action((Array:1))
#17 .../wp-admin/admin.php(259): do_action("toplevel_page_CiviCRM")
#18 {main}
 Sorry, due to an error, we are unable to fulfill your request at the moment. You may want to contact your administrator or service provider with more details about what action you were performing when this occurred.
DB Error: already exists
+++++++++++++++++
The ConfigAndLog file shows:
Jun 04 12:27:24  [error] $Fatal Error Details = Array
(
    [callback] => Array
        (
            [0] => CRM_Core_Error
            [1] => exceptionHandler
        )

    [code] => -5
    [message] => DB Error: already exists
    [mode] => 16
    [debug_info] => INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `data_type` , `html_type` , `is_required` , `is_searchable` , `is_search_range` , `weight` , `is_active` , `is_view` , `text_length` , `column_name` , `option_group_id` , `in_selector` ) VALUES ( 12 , 'electoral_level' , 'Level' , 'String' , 'Select' ,  0 ,  1 ,  0 ,  1 ,  1 ,  0 ,  128 , 'electoral_districts_level' ,  95 ,  1 )  [nativecode=1062 ** Duplicate entry 'Level-12' for key 'UI_label_custom_group_id']
    [type] => DB_Error
    [user_info] => INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `data_type` , `html_type` , `is_required` , `is_searchable` , `is_search_range` , `weight` , `is_active` , `is_view` , `text_length` , `column_name` , `option_group_id` , `in_selector` ) VALUES ( 12 , 'electoral_level' , 'Level' , 'String' , 'Select' ,  0 ,  1 ,  0 ,  1 ,  1 ,  0 ,  128 , 'electoral_districts_level' ,  95 ,  1 )  [nativecode=1062 ** Duplicate entry 'Level-12' for key 'UI_label_custom_group_id']
    [to_string] => [db_error: message="DB Error: already exists" code=-5 mode=callback callback=CRM_Core_Error::exceptionHandler prefix="" info="INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `data_type` , `html_type` , `is_required` , `is_searchable` , `is_search_range` , `weight` , `is_active` , `is_view` , `text_length` , `column_name` , `option_group_id` , `in_selector` ) VALUES ( 12 , 'electoral_level' , 'Level' , 'String' , 'Select' ,  0 ,  1 ,  0 ,  1 ,  1 ,  0 ,  128 , 'electoral_districts_level' ,  95 ,  1 )  [nativecode=1062 ** Duplicate entry 'Level-12' for key 'UI_label_custom_group_id']"]
)


Jun 04 12:27:24  [debug] $backTrace = #0 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Error.php(942): CRM_Core_Error::backtrace("backTrace", TRUE)
#1 .../wp-content/plugins/civicrm/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(922): CRM_Core_Error::exceptionHandler(Object(DB_Error))
#2 .../wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB.php(997): PEAR_Error->__construct("DB Error: already exists", -5, 16, (Array:2), "INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `d...")
#3 .../wp-content/plugins/civicrm/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(575): DB_Error->__construct(-5, 16, (Array:2), "INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `d...")
#4 .../wp-content/plugins/civicrm/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(223): PEAR::_raiseError(Object(DB_mysqli), NULL, -5, 16, (Array:2), "INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `d...", "DB_Error", TRUE)
#5 .../wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/common.php(1928): PEAR->__call("raiseError", (Array:7))
#6 .../wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/mysqli.php(936): DB_common->raiseError(-5, NULL, NULL, "INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `d...", "1062 ** Duplicate entry 'Level-12' for key 'UI_label_custom_group_id'")
#7 .../wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/mysqli.php(406): DB_mysqli->mysqliRaiseError()
#8 .../wp-content/plugins/civicrm/civicrm/vendor/pear/db/DB/common.php(1234): DB_mysqli->simpleQuery("INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `d...")
#9 .../wp-content/plugins/civicrm/civicrm/packages/DB/DataObject.php(2696): DB_common->query("INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `d...")
#10 .../wp-content/plugins/civicrm/civicrm/packages/DB/DataObject.php(1245): DB_DataObject->_query("INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `d...")
#11 .../wp-content/plugins/civicrm/civicrm/CRM/Core/DAO.php(639): DB_DataObject->insert()
#12 .../wp-content/plugins/civicrm/civicrm/CRM/Core/BAO/CustomField.php(2040): CRM_Core_DAO->save()
#13 .../wp-content/plugins/civicrm/civicrm/CRM/Core/BAO/CustomField.php(89): CRM_Core_BAO_CustomField::createCustomFieldRecord((Array:19))
#14 .../wp-content/plugins/civicrm/civicrm/Civi/Api4/Generic/Traits/DAOActionTrait.php(140): CRM_Core_BAO_CustomField::create((Array:18))
#15 .../wp-content/plugins/civicrm/civicrm/Civi/Api4/Generic/DAOSaveAction.php(41): Civi\Api4\Generic\DAOSaveAction->writeObjects((Array:7))
#16 .../wp-content/plugins/civicrm/civicrm/Civi/Api4/Provider/ActionObjectProvider.php(68): Civi\Api4\Generic\DAOSaveAction->_run(Object(Civi\Api4\Generic\Result))
#17 .../wp-content/plugins/civicrm/civicrm/Civi/API/Kernel.php(149): Civi\Api4\Provider\ActionObjectProvider->invoke(Object(Civi\Api4\Generic\DAOSaveAction))
#18 .../wp-content/plugins/civicrm/civicrm/Civi/Api4/Generic/AbstractAction.php(239): Civi\API\Kernel->runRequest(Object(Civi\Api4\Generic\DAOSaveAction))
#19 .../wp-content/plugins/civicrm/civicrm/CRM/Utils/Migrate/Import.php(355): Civi\Api4\Generic\AbstractAction->execute()
#20 .../wp-content/plugins/civicrm/civicrm/CRM/Utils/Migrate/Import.php(66): CRM_Utils_Migrate_Import->customFields(Object(SimpleXMLElement), (Array:2))
#21 .../wp-content/plugins/civicrm/civicrm/CRM/Utils/Migrate/Import.php(43): CRM_Utils_Migrate_Import->runXmlElement(Object(SimpleXMLElement))
#22 .../wp-content/plugins/civicrm/civicrm/ext/com.jlacey.electoral-master/CRM/Electoral/Upgrader/Base.php(96): CRM_Utils_Migrate_Import->run(".../wp-content/plugins/civicrm/civicrm/ext/com.jlacey.elec...")
#23 .../wp-content/plugins/civicrm/civicrm/ext/com.jlacey.electoral-master/CRM/Electoral/Upgrader/Base.php(253): CRM_Electoral_Upgrader_Base::executeCustomDataFileByAbsPath(".../wp-content/plugins/civicrm/civicrm/ext/com.jlacey.elec...")
#24 .../wp-content/plugins/civicrm/civicrm/ext/com.jlacey.electoral-master/electoral.civix.php(50): CRM_Electoral_Upgrader_Base->onInstall()
#25 .../wp-content/plugins/civicrm/civicrm/ext/com.jlacey.electoral-master/electoral.php(62): _electoral_civix_civicrm_install()
#26 .../wp-content/plugins/civicrm/civicrm/CRM/Extension/Manager/Module.php(60): electoral_civicrm_install()
#27 .../wp-content/plugins/civicrm/civicrm/CRM/Extension/Manager/Module.php(32): CRM_Extension_Manager_Module->callHook(Object(CRM_Extension_Info), "install")
#28 .../wp-content/plugins/civicrm/civicrm/CRM/Extension/Manager.php(293): CRM_Extension_Manager_Module->onPreInstall(Object(CRM_Extension_Info))
#29 .../wp-content/plugins/civicrm/civicrm/api/v3/Extension.php(42): CRM_Extension_Manager->install((Array:1))
#30 .../wp-content/plugins/civicrm/civicrm/Civi/API/Provider/MagicFunctionProvider.php(89): civicrm_api3_extension_install((Array:2))
#31 .../wp-content/plugins/civicrm/civicrm/Civi/API/Kernel.php(149): Civi\API\Provider\MagicFunctionProvider->invoke((Array:8))
#32 .../wp-content/plugins/civicrm/civicrm/Civi/API/Kernel.php(81): Civi\API\Kernel->runRequest((Array:8))
#33 .../wp-content/plugins/civicrm/civicrm/api/api.php(131): Civi\API\Kernel->runSafe("Extension", "install", (Array:2))
#34 .../wp-content/plugins/civicrm/civicrm/CRM/Admin/Form/Extensions.php(173): civicrm_api3("Extension", "install", (Array:2))
#35 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Form.php(526): CRM_Admin_Form_Extensions->postProcess()
#36 .../wp-content/plugins/civicrm/civicrm/CRM/Core/StateMachine.php(144): CRM_Core_Form->mainProcess()
#37 .../wp-content/plugins/civicrm/civicrm/CRM/Core/QuickForm/Action/Next.php(43): CRM_Core_StateMachine->perform(Object(CRM_Admin_Form_Extensions), "next", "Next")
#38 .../wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Controller.php(203): CRM_Core_QuickForm_Action_Next->perform(Object(CRM_Admin_Form_Extensions), "next")
#39 .../wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Page.php(103): HTML_QuickForm_Controller->handle(Object(CRM_Admin_Form_Extensions), "next")
#40 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Controller.php(347): HTML_QuickForm_Page->handle("next")
#41 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Page/Basic.php(395): CRM_Core_Controller->run()
#42 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Page/Basic.php(140): CRM_Core_Page_Basic->edit(1, NULL)
#43 .../wp-content/plugins/civicrm/civicrm/CRM/Admin/Page/Extensions.php(105): CRM_Core_Page_Basic->run()
#44 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(313): CRM_Admin_Page_Extensions->run((Array:3), NULL)
#45 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(69): CRM_Core_Invoke::runItem((Array:15))
#46 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(36): CRM_Core_Invoke::_invoke((Array:3))
#47 .../wp-content/plugins/civicrm/civicrm.php(1172): CRM_Core_Invoke::invoke((Array:3))
#48 .../wp-includes/class-wp-hook.php(292): CiviCRM_For_WordPress->invoke("")
#49 .../wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters("", (Array:1))
#50 .../wp-includes/plugin.php(484): WP_Hook->do_action((Array:1))
#51 .../wp-admin/admin.php(259): do_action("toplevel_page_CiviCRM")
#52 {main}


Jun 04 12:27:24  [error] 
$Fatal Error Details = array:3 [
  "message" => "DB Error: already exists"
  "code" => null
  "exception" => CiviCRM_API3_Exception {#8727
    -extraParams: array:6 [
      "error_code" => "already exists"
      "sql" => "INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `data_type` , `html_type` , `is_required` , `is_searchable` , `is_search_range` , `weight` , `is_active` , `is_view` , `text_length` , `column_name` , `option_group_id` , `in_selector` ) VALUES ( 12 , 'electoral_level' , 'Level' , 'String' , 'Select' ,  0 ,  1 ,  0 ,  1 ,  1 ,  0 ,  128 , 'electoral_districts_level' ,  95 ,  1 )  [nativecode=1062 ** Duplicate entry 'Level-12' for key 'UI_label_custom_group_id']"
      "tip" => "add debug=1 to your API call to have more info about the error"
      "is_error" => 1
      "error_message" => "DB Error: already exists"
      "debug_information" => "INSERT INTO `civicrm_custom_field` (`custom_group_id` , `name` , `label` , `data_type` , `html_type` , `is_required` , `is_searchable` , `is_search_range` , `weight` , `is_active` , `is_view` , `text_length` , `column_name` , `option_group_id` , `in_selector` ) VALUES ( 12 , 'electoral_level' , 'Level' , 'String' , 'Select' ,  0 ,  1 ,  0 ,  1 ,  1 ,  0 ,  128 , 'electoral_districts_level' ,  95 ,  1 )  [nativecode=1062 ** Duplicate entry 'Level-12' for key 'UI_label_custom_group_id']"
    ]
    #message: "DB Error: already exists"
    #code: 0
    #file: ".../wp-content/plugins/civicrm/civicrm/api/api.php"
    #line: 133
    trace: {
      .../wp-content/plugins/civicrm/civicrm/api/api.php:133 {
        â€º if (is_array($result) && !empty($result['is_error'])) {
        â€º   throw new CiviCRM_API3_Exception($result['error_message'], CRM_Utils_Array::value('error_code', $result, 'undefined'), $result);
        â€º }
      }
      .../wp-content/plugins/civicrm/civicrm/CRM/Admin/Form/Extensions.php:173 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/Form.php:526 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/StateMachine.php:144 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/QuickForm/Action/Next.php:43 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Controller.php:203 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/packages/HTML/QuickForm/Page.php:103 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/Controller.php:347 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/Page/Basic.php:395 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/Page/Basic.php:140 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Admin/Page/Extensions.php:105 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php:313 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php:69 { â€¦}
      .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php:36 { â€¦}
      .../wp-content/plugins/civicrm/civicrm.php:1172 { â€¦}
      .../wp-includes/class-wp-hook.php:292 { â€¦}
      .../wp-includes/class-wp-hook.php:316 { â€¦}
      .../wp-includes/plugin.php:484 { â€¦}
      .../wp-admin/admin.php:259 { â€¦}
    }
  }
]


Jun 04 12:27:24  [debug] $backTrace = #0 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Error.php(433): CRM_Core_Error::backtrace("backTrace", TRUE)
#1 .../wp-content/plugins/civicrm/civicrm/CRM/Core/Invoke.php(39): CRM_Core_Error::handleUnhandledException(Object(CiviCRM_API3_Exception))
#2 .../wp-content/plugins/civicrm/civicrm.php(1172): CRM_Core_Invoke::invoke((Array:3))
#3 .../wp-includes/class-wp-hook.php(292): CiviCRM_For_WordPress->invoke("")
#4 .../wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters("", (Array:1))
#5 .../wp-includes/plugin.php(484): WP_Hook->do_action((Array:1))
#6 .../wp-admin/admin.php(259): do_action("toplevel_page_CiviCRM")
#7 {main}

**Comment by josephlacey** [comment]
Hi @jimcrist,

You could remove this file from the extension and retry the install,

https://github.com/josephlacey/com.jlacey.electoral/blob/master/xml/auto_install.xml

This assume you have all the custom data fields and option lists in place and would skip the creation of those.

Joseph

**Comment by privaterra** [comment]
When I’ve had issues like this, it likely indicates that the extension has been removed - but the related database tables - have not been deleted/removed.

Where the extension has been installed previously, disable and uninstall the extension. Update the code, and instead of enabling the code via the CiviCRM system settings, enable it by manually enabling the extension in the database using a tool like phpMyAdmin


The list of installed & enabled extensions are in the table -  civicrm_extension .  If there’s no entry for the “Electoral API”, then you’ll have to create it..

ie. insert a new line, here’s the format I used to add/enable the extension.

ID - next available ID (last one +1)
type - module
full_name - com.jlacey.electoral
name - Electoral API
label - Electoral API
file - electoral
schema_version - (set to null)
is_active - 1

Save (go in phpmyadmin)

return to CiviCRM - the extension should now appear as enabled, and hopefully works….

regards

Robert


--
Robert Guerra
Cel/Tel +1 416 893 0377
Twitter: twitter.com/netfreedom
Email: ***@***.***
